collector/mdadm.go

```diff
@@ -0,0 +1,200 @@
+// +build !nomdadm
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/log"
+)
+
+var statusfile = "/proc/mdstat"
+
+type mdstatus struct {
+	mdName       string
+	isActive     bool
+	disksActive  int64
+	disksTotal   int64
+	blocksTotal  int64
+	blocksSynced int64
+}
+
+type mdadmCollector struct {
+	isActive     *prometheus.GaugeVec
+	disksActive  *prometheus.GaugeVec
+	disksTotal   *prometheus.GaugeVec
+	blocksTotal  *prometheus.GaugeVec
+	blocksSynced *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["mdadm"] = NewMdadmCollector
+}
+
+func evalStatusline(statusline string) (active, total, size int64) {
+
+	first_split := strings.Split(statusline, "[")
+	second_split := strings.Split(first_split[1], "]")
+
+	sizeDirty := strings.Split(first_split[0], "blocks")[0]
+	sizeStripped := strings.Replace(sizeDirty, " ", "", -1)
+
+	numberOfDrives := strings.Split(second_split[0], "/")
+
+	total, _ = strconv.ParseInt(numberOfDrives[0], 10, 64)
```

Comment:
The way your code is currently written (`parseMdstat` being called before any metrics are put into the channel in `Update()`), this would cause no metrics to be outputted from this collector at all in case of error. I'd argue that's probably the best thing to do. Errors here should be the really exceptional case, so let's break the collector completely in that case.